assembly/src/compile.rs

```diff
@@ -0,0 +1,332 @@
+use alloc::{
+    borrow::Cow,
+    boxed::Box,
+    string::{String, ToString},
+    sync::Arc,
+    vec::Vec,
+};
+
+use crate::{
+    ast::{Module, ModuleKind},
+    diagnostics::{
+        Diagnostic, IntoDiagnostic, NamedSource, Report, SourceCode, SourceFile, WrapErr,
+    },
+    library::{LibraryNamespace, LibraryPath},
+};
+
+// COMPILE OPTIONS
+// ================================================================================================
+
+/// The set of options which can be used to control the behavior of the [Compile] trait.
+#[derive(Debug, Clone)]
+pub struct Options {
+    /// The kind of [Module] to compile.
+    ///
+    /// The default kind is executable.
+    pub kind: ModuleKind,
+    /// The namespace to apply to the compiled [Module].
+    ///
+    /// If unset, the namespace will be derived from the [ModuleKind].
+    pub namespace: Option<LibraryNamespace>,
+    /// The name to give the compiled [Module]
+    ///
+    /// This option overrides `namespace`.
+    ///
+    /// If unset, and there is no name associated with the item being compiled (e.g. a file path)
+    /// then the path will consist of just a namespace; using the value of `namespace` if provided,
+    /// or deriving one from `kind`.
+    pub path: Option<LibraryPath>,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            kind: ModuleKind::Executable,
+            namespace: None,
+            path: None,
+        }
+    }
+}
+impl Options {
+    /// Configure a set of [Options] to compile a [Module] with the given `kind` and `path`.
+    ///
+    /// This is primarily useful when compiling a module from source code that has no meaningful
+    /// [LibraryPath] associated with it, such as when compiling from a `str`. This will override
+    /// the default name derived from the given [ModuleKind].
+    pub fn new<P, E>(kind: ModuleKind, path: P) -> Result<Self, E>
+    where
+        P: TryInto<LibraryPath, Error = E>,
+    {
+        let path = path.try_into()?;
+        Ok(Self {
+            kind,
+            namespace: None,
+            path: Some(path),
+        })
+    }
+
+    /// Get the default [Options] for compiling a library module.
+    pub fn for_library() -> Self {
+        Self {
+            kind: ModuleKind::Library,
+            ..Default::default()
+        }
+    }
+
+    /// Get the default [Options] for compiling a kernel module.
+    pub fn for_kernel() -> Self {
+        Self {
+            kind: ModuleKind::Kernel,
+            ..Default::default()
+        }
+    }
+}
+
+/// This error occurs when attempting to use a [Module] in a context which is invalid for that
+/// module.
+///
+/// For example, using an executable module as a library.
+#[derive(Debug, thiserror::Error, Diagnostic)]
+#[error("compilation failed: expected a {expected} module, but got a {actual} module")]
+#[diagnostic()]
+pub struct ModuleKindMismatchError {
+    expected: ModuleKind,
+    actual: ModuleKind,
+}
```

Comment:
Is this error used anywhere? I could find any reference to it.

Reply:
Ah indeed, it was also made redundant by those other changes that made the `namespace` option no longer necessary. I'll revisit the affected code and clean that stuff up.